File: 08_05_2024_Root_to_Leaf_Paths.java
class Solution {
    public List<List<Integer>> Paths(Node root) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();
        solve(root, curr, ans);
        return ans;
    }

    private void solve(Node root, List<Integer> curr, List<List<Integer>> ans) {
        if (root == null) {
            return;
        }
        curr.add(root.data);
        if (root.left == null && root.right == null) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        solve(root.left, curr, ans);
        solve(root.right, curr, ans);
        curr.remove(curr.size() - 1);
    }
}

File: 08_05_2024_Root_to_Leaf_Paths.py
class Solution:
    def Paths(self, root):
        def solve(node, curr, ans):
            if not node:
                return
            curr.append(node.data)
            if not node.left and not node.right:
                ans.append(curr[:])
                return
            solve(node.left, curr, ans)
            solve(node.right, curr, ans)
            curr.pop()

        ans = []
        curr = []
        solve(root, curr, ans)
        return ans

